src/java.base/share/classes/jdk/internal/math/FloatingDecimal.java

```diff
@@ -1826,726 +1825,603 @@ private static BinaryToASCIIConverter getBinaryToASCIIConverter(float f) {
+        while (i < len && in.charAt(i) <= ' ') {
+            ++i;
+        }
+        return i;
+    }
 
-                    return new PreparedASCIIToBinaryBuffer(value, fValue);
-                }
-            }
+    /*
+     * Attempts to scan sub and opt trailing whitespaces, starting at index i.
+     * The optional whitespaces must be at the end of in.
+     */
+    private static void scanSymbolic(String in, int i, String sub) {
+        int high = i + sub.length();  // might overflow, checked below
+        check(in, i <= high && high <= in.length()
+                        && in.indexOf(sub, i, high) == i
+                        && skipWhitespaces(in, high) == in.length());
     }
 
-    /**
-     * Returns <code>s</code> with any leading zeros removed.
+    /*
+     * Returns i if this is the first time the scanner detects a point.
+     * Throws otherwise.
      */
-    static String stripLeadingZeros(String s) {
-        if(!s.isEmpty() && s.charAt(0)=='0') {
-            for(int i=1; i<s.length(); i++) {
-                if(s.charAt(i)!='0') {
-                    return s.substring(i);
-                }
-            }
-            return "";
+    private static int checkMultiplePoints(int pt, int i) {
+        if (pt != 0) {
+            throw new NumberFormatException("multiple points");
         }
-        return s;
+        return i;
     }
 
-    /**
-     * Extracts a hexadecimal digit from position <code>position</code>
-     * of string <code>s</code>.
-     */
-    static int getHexDigit(String s, int position) {
-        int value = Character.digit(s.charAt(position), 16);
-        if (value <= -1 || value >= 16) {
-            throw new AssertionError("Unexpected failure of digit conversion of " +
-                                     s.charAt(position));
+    private static final int MAX_OUT = 1_000;
+    private static final String OMITTED = " ... ";
+
+    private static void check(String in, boolean b) {
+        if (!b) {
+            int len = in.length();
+            if (len > MAX_OUT) {  // discard middle chars to achieve a length of MAX_OUT
+                int halfish = (MAX_OUT - OMITTED.length()) / 2;
+                in = in.substring(0, halfish)
+                        + OMITTED
+                        + in.substring(len - (MAX_OUT - OMITTED.length() - halfish));
+            }
+            throw new NumberFormatException("For input string: \"" + in + "\"");
         }
-        return value;
     }
+
+    /*
+     * These constants are used to indicate the IEEE binary floating-point format
+     * as an index (ix) to some methods and static arrays in this class.
+     */
+    private static final int BINARY_16_IX = 0;
+    private static final int BINARY_32_IX = 1;
+    private static final int BINARY_64_IX = 2;
+//    private static final int BINARY_128_IX = 3;
+//    private static final int BINARY_256_IX = 4;
+
+    /* The precision of the format. */
+    private static final int[] P = {
```

Comment:
What about organizing these constants as as a record - something like this:

```
record FloatFmt(int ix, int p, int epMin, int epMax, int w, int eMin, int eMax, int qMin, int hexCount) {

    public static final FloatFmt BINARY_16 = build(0, 11, -8, 6);
    public static final FloatFmt BINARY_32 = build(1, 24, -46, 40);
    public static final FloatFmt BINARY_64 = build(2, 53, -324, 310);
    public static final FloatFmt BINARY_128 = build(3, 113, -4_966, 4934);
    public static final FloatFmt BINARY_256 = build(4, 237, -78_985, 78_915);

    private static FloatFmt build(int ix, int p, int epMin, int epMax) {
        int w = (1 << 4 + ix) - p;
        int eMin = (-1 << w - 1) + 2;
        int eMax = (1 << w - 1) - 1;
        int qMin = eMin - p + 1;
        int hexCount = p / 4 + 2;

        return new FloatFmt(ix, p, epMin, epMax, w, eMin, eMax, qMin, hexCount);
    }
}

```

Methods could then be parameterized with the static instances rather than integers. I believe these values would inline well as constants.